### PR TITLE
Avoid duplicate diagonalizing_gates

### PR DESCRIPTION
**Pull request by lillian542**
**Context:**
Currently, `diagonalizing_gates` for a tape just iterates over all the observables and adds their diagonalizing gates to the list. This means that if the same observable occurs more than once, it's diagonalizing gates are included more than once.

```
>>> tape = qml.tape.QuantumScript([], [qml.expval(X(0)), qml.var(X(0))])
>>> tape.diagonalizing_gates
[Hadamard(wires=[0]), Hadamard(wires=[0])]
```

**Description of the Change:**
We get diagonalizing gates from a set of the base observables instead of `self.observables`. The bulk of this change is a `_get_base_obs` function that iterates over observables on the tape and gets the individual elements that are being diagonalized.

**Benefits:**
```
>>> tape = qml.tape.QuantumScript([], [qml.expval(X(0)), qml.var(X(0))])
>>> tape.diagonalizing_gates
[Hadamard(wires=[0])]
```

------------------

**Open question:**
One slightly unintuitive thing here is how to handle cases with `CompositeOp` observable where an observable is duplicated, but in an observable that can't be divided because its terms don't commute. For example, currently in this PR for:

```
tape.observables = [qml.X(0), qml.X(0)+qml.Y(0)]
```

Here we include `X(0).diagonalizing_gates()` and the `QubitUnitary` that diagonalized `qml.X(0)+qml.Y(0)`. Since the sum can't be diagonalized as a single observable we treat is as a completely different observable, the way we would:

```
tape.observables = [qml.X(0), qml.Y(0)]
```
In contrast, if the sum doesn't have overlapping wires, and thus would be diagonalized one term at a time instead of as a single observable:

```
tape.observables = [qml.X(0), qml.X(0)+qml.Y(1)]
```

we end up with `X(0).diagonalizing_gates() + Y(1).diagonalizing_gates()` 

This feels more consistent now that we have a diagonalizing gate for X(0)+Y(0), but it doesn't reflect how we currently diagonalize if using the `diagonalize_measurements` transform, which assumes we've already applied `split_non_commuting` and that X(0) and Y(0) are therefore on separate tapes. But if people want that behaviour, they should apply `split_non_commuting` to the tape...

Essentially, is it strange that a tape with a single observable `[qml.X(0) + qml.Y(1)]` has the same diagonalizing gates as a tape with observables `[qml.X(0), qml.Y(1)]`, but a tape with `[qml.X(0) + qml.Y(0)]` and `[qml.X(0), qml.Y(0)]` have different diagonalizing gates? 

I _think_ this is the most consistent choice.

[sc-64695]

### CODE REVIEW

**Review by albi3ro — COMMENTED** [review]
Can we add some extra information and examples to the docstring, potentially with a warning box?

**Comment by lillian542** [comment]
> If your tape contains measurements of both X(0) and X(0) + Y(0), which don't commute, no sequence of gates is going to be able to diagonalize for both of these measurements. In this case, tape.diagonalizing_gates could produce garbage for all we care.

I don't disagree, but the decision when it was discussed was that we are happy with it acting like this when there are non-commuting measurements - see discussion linked on the shortcut story 🤷‍♀️ But I would think in all the cases where this property matters, you should be looking at tapes that have already been split into individual executions anyway.

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -128,6 +133,10 @@
 * `Operator.expand` is now removed. Use `qml.tape.QuantumScript(op.deocomposition())` instead.
   [(#6227)](https://github.com/PennyLaneAI/pennylane/pull/6227)
 
+* When an observable is repeated on a tape, `tape.diagonalizing_gates` no longer returns the
```
[the remark this answers is not in the document]
Comment (lillian542) [reply doc/releases/changelog-dev.md]: removed

**Comment by lillian542** [comment]
I added a more extensive docstring for `diagonalizing_gates`, including a warning about non-commuting observables

**Review by astralcai — APPROVED** [review]
LGTM

**Review by albi3ro — APPROVED** [review]
👍  thanks for the docs

**Pull request merged**